### ISSUE TITLE
Adapt to new numpy.random interface

### ISSUE THREAD

**Issue by HDembinski** [dsaxton/resample#5]
I am looking at the function interfaces in bootstrap.py. I would like to discuss how those could be enhanced to use the new numpy.random interface, where you don't pass a seed to the client, but the seeded generator instance.

1) Are you ok in requiring the minimum numpy version where this generator interface was added, or do you need to support older numpy versions?

2) The backward compatible way would be to add a new keyword `rng=None` to each method. If it is `None`, then use the default random number generator. Otherwise, use the passed `numpy.random.Generator` instance. For `bootstrap*`, I would deprecate the keyword `random_state` and eventually replace it by `rng`.

**Comment by dsaxton** [comment]
> 1. Are you ok in requiring the minimum numpy version where this generator interface was added, or do you need to support older numpy versions?

I think bumping the version makes sense. IIRC the minimum version should be something like 1.17 in that case?

> 2. The backward compatible way would be to add a new keyword `rng=None` to each method. If it is `None`, then use the default random number generator. Otherwise, use the passed `numpy.random.Generator` instance. For `bootstrap*`, I would deprecate the keyword `random_state` and eventually replace it by `rng`.

TBH I don't think we need to worry about backward compatibility too much, and while the name random_state is somewhat of a convention in packages like sklearn and pandas, I'd be okay breaking with that to use a more appropriate one.

**Comment by HDembinski** [comment]
> I think bumping the version makes sense. IIRC the minimum version should be something like 1.17 in that case?

Ok, nice, that's the easiest way to code this. Yes, [the generator interface was added in 1.17](https://docs.scipy.org/doc/numpy/release.html).

> TBH I don't think we need to worry about backward compatibility too much, and while the name random_state is somewhat of a convention in packages like sklearn and pandas, I'd be okay breaking with that to use a more appropriate one.

I am not super familiar with pandas and sklearn. Following the conventions of other popular packages is good, although I suppose that those do not support the new numpy generator interface yet, either.

**Comment by HDembinski** [comment]
The main question then is: what kind of keyword to the others use to pass the random number generator instance? I didn't check, I just suggested `rng` because it seemed appropriate.

**Comment by dsaxton** [comment]
> The main question then is: what kind of keyword do the others use to pass the random number generator instance? I didn't check, I just suggested `rng` because it seemed appropriate.

I actually like the `rng` keyword, it expresses what the argument is and it's easy to type.

**Comment by HDembinski** [comment]
I will also address this in #28

**Comment by HDembinski** [comment]
After looking the scipy code and how they handle the transition, I think we should stick the the keyword `random_state` to pass the random number generator instance. That's how scipy does it and we should be consistent, especially if we want to merge this code into scipy at some point.
